README.md
# flutter_scroll_gallery

A Flutter package that helps you to create a zoomable carousel with scrollable thumbnail at the bottom

Inspired from [flutter_image_carousel](https://github.com/theobouwman/flutter_image_carousel) and [ScrollGalleryView](https://github.com/VEINHORN/ScrollGalleryView)

## Installing

```yaml
dependencies:
  flutter:
    sdk: flutter
  flutter_scroll_gallery: ^0.1.0
```

```dart
import 'package:flutter_scroll_gallery/flutter_scroll_gallery.dart';
```

## Usage

```dart
new ScrollGallery(
  <ImageProvider>[
    new NetworkImage("https://flutter.io/images/homepage/header-illustration.png"),
    new NetworkImage("https://i.ytimg.com/vi/fq4N0hgOWzU/maxresdefault.jpg"),
    new NetworkImage("https://cdn-images-1.medium.com/max/1219/1*TFZQzyVAHLVXI_wNreokGA.png"),
    new NetworkImage("https://cdn-images-1.medium.com/max/744/1*L2IaQThqx4lzz3G1m-e07Q.png"),
    // new NetworkImage("https://flutter.io/images/homepage/header-illustration.png"),
    // new NetworkImage("https://i.ytimg.com/vi/fq4N0hgOWzU/maxresdefault.jpg"),
    // new NetworkImage("https://cdn-images-1.medium.com/max/1219/1*TFZQzyVAHLVXI_wNreokGA.png"),
    // new NetworkImage("https://cdn-images-1.medium.com/max/744/1*L2IaQThqx4lzz3G1m-e07Q.png"),
    // new NetworkImage("https://flutter.io/images/homepage/header-illustration.png"),
    // new NetworkImage("https://i.ytimg.com/vi/fq4N0hgOWzU/maxresdefault.jpg"),
    // new NetworkImage("https://cdn-images-1.medium.com/max/1219/1*TFZQzyVAHLVXI_wNreokGA.png"),
    // new NetworkImage("https://cdn-images-1.medium.com/max/744/1*L2IaQThqx4lzz3G1m-e07Q.png")
  ],
  interval: new Duration(seconds: 3),)
);
```

To disable auto-play, do not set the **interval**

## Showcase

![Showcase](showcase.gif)

CHANGELOG.md
## [1.0.1]

* Fix late _timer was not initialized but called

## [1.0.0]

* Migraate to Nullsafety

## [0.1.0]

* Improve code style and performance
* Support zoom using PhotoView
* Support change background color

## [0.0.2]

* Upgrade to Dart 2

## [0.0.1]

* Initial release


lib/src/scroll_gallery.dart
import 'dart:async';
import 'package:flutter/material.dart';
import 'package:photo_view/photo_view.dart';

typedef void OnPageChange(int index);

class ScrollGallery extends StatefulWidget {
  final double height;
  final double thumbnailSize;
  final List<ImageProvider> imageProviders;
  final BoxFit fit;
  final Duration? interval;
  final Color borderColor;
  final Color backgroundColor;
  final bool zoomable;
  final int initialIndex;
  final OnPageChange? onPageChange;

  ScrollGallery(
    this.imageProviders, {
    this.height = double.infinity,
    this.thumbnailSize = 48.0,
    this.borderColor = Colors.red,
    this.backgroundColor = Colors.black,
    this.zoomable = true,
    this.fit = BoxFit.contain,
    this.interval,
    this.initialIndex = 0,
    this.onPageChange,
  });

  @override
  _ScrollGalleryState createState() => _ScrollGalleryState();
}

class _ScrollGalleryState extends State<ScrollGallery>
    with SingleTickerProviderStateMixin {
  late final ScrollController _scrollController;
  late final PageController _pageController;
  Timer? _timer;
  int _currentIndex = 0;
  bool _reverse = false;
  bool _lock = false;

  @override
  void initState() {
    _scrollController = new ScrollController();
    _pageController = new PageController(initialPage: widget.initialIndex);
    _currentIndex = widget.initialIndex;
    final interval = widget.interval;
    if (interval != null && widget.imageProviders.length > 1) {
      _timer = new Timer.periodic(interval, (_) {
        if (_lock) {
          return;
        }

        if (_currentIndex == widget.imageProviders.length - 1) {
          _reverse = true;
        }
        if (_currentIndex == 0) {
          _reverse = false;
        }

        if (_reverse) {
          _pageController.previousPage(
            duration: const Duration(milliseconds: 500),
            curve: Curves.ease,
          );
        } else {
          _pageController.nextPage(
            duration: const Duration(milliseconds: 500),
            curve: Curves.ease,
          );
        }
      });
    }

    super.initState();
  }

  @override
  void dispose() {
    _timer?.cancel();
    _scrollController.dispose();
    _pageController.dispose();
    super.dispose();
  }

  void _onPageChanged(int index) {
    widget.onPageChange?.call(index);
    setState(() {
      _currentIndex = index;
      double itemSize = widget.thumbnailSize + 8.0;
      _scrollController.animateTo(
        itemSize * index / 2,
        duration: const Duration(milliseconds: 200),
        curve: Curves.ease,
      );
    });
  }

  Widget _zoomableImage(image) {
    return new PhotoView(
      backgroundDecoration: BoxDecoration(color: widget.backgroundColor),
      imageProvider: image,
      minScale: PhotoViewComputedScale.contained,
      scaleStateChangedCallback: (PhotoViewScaleState state) {
        setState(() {
          _lock = state != PhotoViewScaleState.initial;
        });
      },
    );
  }

  Widget _notZoomableImage(image) {
    return new Image(image: image, fit: widget.fit);
  }

  Widget _buildImagePageView() {
    return Expanded(
      child: new PageView(
        physics: _lock ? NeverScrollableScrollPhysics() : null,
        onPageChanged: _onPageChanged,
        controller: _pageController,
        children: widget.imageProviders.map((image) {
          return (widget.zoomable
              ? _zoomableImage(image)
              : _notZoomableImage(image));
        }).toList(),
      ),
    );
  }

  void _selectImage(int index) {
    setState(() {
      _pageController.animateToPage(
        index,
        duration: const Duration(milliseconds: 500),
        curve: Curves.ease,
      );
      _lock = false;
    });
  }

  Widget _buildImageThumbnail() {
    return new Container(
      height: widget.thumbnailSize,
      child: new ListView.builder(
        controller: _scrollController,
        itemCount: widget.imageProviders.length,
        scrollDirection: Axis.horizontal,
        itemBuilder: (BuildContext context, int index) {
          var decoration = new BoxDecoration(color: Colors.white);

          if (_currentIndex == index) {
            decoration = new BoxDecoration(
              border: new Border.all(
                color: widget.borderColor,
                width: 2.0,
              ),
              color: Colors.white,
            );
          }

          return new GestureDetector(
            onTap: () {
              _selectImage(index);
            },
            child: new Container(
              decoration: decoration,
              margin: const EdgeInsets.only(left: 8.0),
              child: new Image(
                image: widget.imageProviders[index],
                fit: widget.fit,
                width: widget.thumbnailSize,
                height: widget.thumbnailSize,
              ),
            ),
          );
        },
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      height: widget.height,
      color: widget.backgroundColor,
      child: new Column(
        crossAxisAlignment: CrossAxisAlignment.center,
        children: <Widget>[
          _buildImagePageView(),
          new SizedBox(height: 8.0),
          _buildImageThumbnail(),
          new SizedBox(height: 8.0)
        ],
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_scroll_gallery/flutter_scroll_gallery.dart';

void main() => runApp(new MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      title: 'Flutter Demo',
      theme: new ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or press Run > Flutter Hot Reload in IntelliJ). Notice that the
        // counter didn't reset back to zero; the application is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: new MyImageGallery(),
    );
  }
}

class MyImageGallery extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: new AppBar(
        title: new Text("Flutter Scroll Gallery"),
      ),
      body: new ScrollGallery(<ImageProvider>[
        new NetworkImage("https://flutter.io/images/homepage/header-illustration.png"),
        new NetworkImage("https://i.ytimg.com/vi/fq4N0hgOWzU/maxresdefault.jpg"),
        new NetworkImage("https://cdn-images-1.medium.com/max/1219/1*TFZQzyVAHLVXI_wNreokGA.png"),
        new NetworkImage("https://cdn-images-1.medium.com/max/744/1*L2IaQThqx4lzz3G1m-e07Q.png"),
        // new NetworkImage("https://flutter.io/images/homepage/header-illustration.png"),
        // new NetworkImage("https://i.ytimg.com/vi/fq4N0hgOWzU/maxresdefault.jpg"),
        // new NetworkImage("https://cdn-images-1.medium.com/max/1219/1*TFZQzyVAHLVXI_wNreokGA.png"),
        // new NetworkImage("https://cdn-images-1.medium.com/max/744/1*L2IaQThqx4lzz3G1m-e07Q.png"),
        // new NetworkImage("https://flutter.io/images/homepage/header-illustration.png"),
        // new NetworkImage("https://i.ytimg.com/vi/fq4N0hgOWzU/maxresdefault.jpg"),
        // new NetworkImage("https://cdn-images-1.medium.com/max/1219/1*TFZQzyVAHLVXI_wNreokGA.png"),
        // new NetworkImage("https://cdn-images-1.medium.com/max/744/1*L2IaQThqx4lzz3G1m-e07Q.png")
      ],
      interval: new Duration(seconds: 3),
      backgroundColor: Colors.blue,
      initialIndex: 2,
      onPageChange: (index) {
        print(index);
      },)
    );
  }
}
